File: Unidad_ii_Testing Orientado a Objetos/ejercicio_3/src/main/java/flota/Auto.java
package flota;

//segun el ejercicio 1
public class Auto extends Vehiculo {
    private int numPuertas;

    public Auto(String patente, String marca, int anio, double kmActuales, int numPuertas) {
        super(patente, kmActuales);
        this.numPuertas = numPuertas;
    }

    @Override
    public String asignar(String legajo, double kmEstimados) {
        // Precondiciones del contrato de Vehiculo 
        if (estado != EstadoVehiculo.DISPONIBLE) {
            throw new IllegalStateException("Vehiculo no disponible"); // 
        }
        if (kmEstimados <= 0 || kmEstimados > 500) {
            throw new IllegalArgumentException("km estimados fuera de rango"); 
        }
        
        this.estado = EstadoVehiculo.EN_USO; 
        return "Asignado a " + legajo + " por " + kmEstimados + " km"; 
    }

    @Override
    public void liberar(double kmRecorridos) {
        // Precondiciones del contrato 
        if (estado != EstadoVehiculo.EN_USO) {
            throw new IllegalStateException("El vehiculo no está en uso"); 
        }
        if (kmRecorridos < 0) {
            throw new IllegalArgumentException("km no puede ser negativo"); 
        }
        
        this.kmActuales += kmRecorridos; 
        this.estado = EstadoVehiculo.DISPONIBLE; 
    }
}

File: Unidad_III_Testing de_Integracio_Regresion_Mutacional_Aceptacion_CICD/Ejercicio2/modelo/auto.py
from enum import Enum


class EstadoVehiculo(Enum):
    DISPONIBLE    = 'disponible'
    EN_USO        = 'en_uso'
    MANTENIMIENTO = 'mantenimiento'
    BAJA          = 'baja'


class Auto:
    def __init__(self, patente: str, marca: str, anio: int,
                 km_actuales: float, num_puertas: int):
        if not patente or len(patente) < 6:
            raise ValueError('Patente inválida')
        if km_actuales < 0:
            raise ValueError('km_actuales no puede ser negativo')
        if num_puertas not in (2, 4, 5):
            raise ValueError('num_puertas debe ser 2, 4 o 5')
        self._patente    = patente
        self._marca      = marca
        self._anio       = anio
        self._km         = km_actuales
        self._puertas    = num_puertas
        self._estado     = EstadoVehiculo.DISPONIBLE
        self._conductor  = None

    def asignar(self, legajo_conductor: str, km_estimados: float) -> str:
        """Cambio A aplicado: límite subió de 500 a 800, y se corrigió el >= (bug)."""
        if self._estado != EstadoVehiculo.DISPONIBLE:
            raise PermissionError('Vehículo no disponible')
        if km_estimados <= 0 or km_estimados > 800:
            raise ValueError('km_estimados fuera de rango')
        self._estado    = EstadoVehiculo.EN_USO
        self._conductor = legajo_conductor
        return f'Asignado a {legajo_conductor} por {km_estimados} km'

    def liberar(self, km_recorridos: float) -> None:
        if self._estado != EstadoVehiculo.EN_USO:
            raise PermissionError('El vehículo no está en uso')
        if km_recorridos < 0:
            raise ValueError('km_recorridos no puede ser negativo')
        self._km        += km_recorridos
        self._estado     = EstadoVehiculo.DISPONIBLE
        self._conductor  = None

    def enviar_mantenimiento(self) -> None:
        if self._estado == EstadoVehiculo.EN_USO:
            raise PermissionError('No se puede enviar a mantenimiento en uso')
        self._estado = EstadoVehiculo.MANTENIMIENTO

    def dar_de_baja(self) -> None:
        if self._estado == EstadoVehiculo.EN_USO:
            raise PermissionError('No se puede dar de baja un vehículo en uso')
        self._estado = EstadoVehiculo.BAJA

    def get_estado(self) -> EstadoVehiculo:
        return self._estado

    def get_km(self) -> float:
        return self._km

    def get_patente(self) -> str:
        return self._patente

    def get_conductor(self):
        return self._conductor
